### PR TITLE
Filesystem refactor

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?

This is a refactor of the system/filesystem and system/fsstat metricsets for the purpose of fixing https://github.com/elastic/beats/issues/30590 Also also continuing to depricate the `gosigar` library.

I'm still in the process of testing this, but I'm currently juggling 3 other things so I at least need to get the PR in so I don't loose track of what I'm doing.

## Why is it important?

This fixes a bug, removes code from metricbeat, and removes deprecated gosigar code.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

Fixes https://github.com/elastic/beats/issues/30590

### CODE REVIEW

File: libbeat/metric/system/filesystem/filesystem.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+// Fill out computed stats after the platform-specific code fetches metrics from the OS
+func (fs *FSStat) fillMetrics() {
+	fs.Used.Bytes = fs.Total.SubtractOrNone(fs.Free)
+
+	percTotal := fs.Used.Bytes.ValueOr(0) + fs.Avail.ValueOr(0)
+	if percTotal == 0 {
+		return
+	}
+	// I'm not sure why this does Used + avail instead of total, but I'm too afraid to change it
```
Comment (cmacknz) [inline libbeat/metric/system/filesystem/filesystem.go:109]: Should this comment be on the line above?

File: metricbeat/module/system/filesystem/filesystem_test.go [hunk 2]
```diff
@@ -51,10 +54,10 @@ func TestData(t *testing.T) {
 }
 
 func getConfig() map[string]interface{} {
-	ignoreTypes := append(DefaultIgnoredTypes(resolve.NewTestResolver("")), "fuse.lxcfs", "fuse.gvfsd-fuse", "nsfs", "squashfs")
+	//ignoreTypes := append(fs.DefaultIgnoredTypes(resolve.NewTestResolver("")), "fuse.lxcfs", "fuse.gvfsd-fuse", "nsfs", "squashfs")
```
Comment (cmacknz) [inline metricbeat/module/system/filesystem/filesystem_test.go:57]: Was this meant to be left commented out?

**Review by cmacknz — APPROVED** [review]

**Comment by fearful-symmetry** [comment]
Okay, so once this gets merged, I should be able to fix+merge this PR: https://github.com/elastic/elastic-agent-system-metrics/pull/5

**Comment by fearful-symmetry** [comment]
Okay, so that `mage check` role isn't failing when I run it locally, so...not sure what's going on there.

**Comment by cmacknz** [comment]
Looking at this again it's looked like it has changed a fair bit.

> This fixes a bug, removes code from metricbeat, and removes deprecated gosigar code.

What is the bug that was fixed here? It's hard to spot in all the code? Also you need a CHANGELOG entry for it.

**Comment by fearful-symmetry** [comment]
@cmacknz yup, hence why I asked for another review in slack. The original code wasn't respecting `hostfs`, which was the bug. Changelog incoming...

File: libbeat/opt/opt.go [hunk 3]
```diff
@@ -33,15 +33,15 @@ type Int struct {
 	value  int
 }
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply libbeat/opt/opt.go:34]: Oh, it's already in there. Huh.

**Comment by fearful-symmetry** [comment]
Alright, changlog added, reverted some of the opt changes.

File: metricbeat/module/system/filesystem/_meta/fields.yml [hunk 4]
```diff
@@ -25,6 +25,10 @@
       type: long
       description: >
         Total number of inodes on the system, which will be a combination of files, folders, symlinks, and devices.
+    - name: options
+      type: keyword
+      description: >
+        The options present on the filesystem mount.
```
Comment (cmacknz) [inline metricbeat/module/system/filesystem/_meta/fields.yml:31]: Should this option be mentioned in the CHANGELOG?

**Review by cmacknz — APPROVED** [review]
LGTM again, I can't spot anything problematic.

**Comment by toby-sutor** [comment]
It seems like this made it into version 8.3.1 as per https://github.com/elastic/beats/blob/v8.3.1/CHANGELOG.asciidoc#L34.

**Pull request merged**